src/Sylius/Bundle/ApiBundle/Validator/Constraints/EligibleProductVariantToCartValidator.php

```diff
@@ -14,59 +14,74 @@
 
 use Sylius\Bundle\ApiBundle\Command\Cart\AddItemToCart;
+use Sylius\Bundle\ApiBundle\Command\Cart\ChangeItemQuantityInCart;
 use Sylius\Component\Core\Model\OrderInterface;
+use Sylius\Component\Core\Model\OrderItemInterface;
 use Sylius\Component\Core\Model\ProductInterface;
 use Sylius\Component\Core\Model\ProductVariantInterface;
 use Sylius\Component\Core\Repository\OrderRepositoryInterface;
 use Sylius\Component\Inventory\Checker\AvailabilityCheckerInterface;
 use Sylius\Component\Product\Repository\ProductVariantRepositoryInterface;
+use Sylius\Component\Resource\Repository\RepositoryInterface;
 use Symfony\Component\Validator\Constraint;
 use Symfony\Component\Validator\ConstraintValidator;
 use Webmozart\Assert\Assert;
 
 /** @experimental */
-final class AddingEligibleProductVariantToCartValidator extends ConstraintValidator
+final class EligibleProductVariantToCartValidator extends ConstraintValidator
 {
     /** @var ProductVariantRepositoryInterface */
     private $productVariantRepository;
 
+    /** @var RepositoryInterface */
+    private $orderItemRepository;
+
     /** @var OrderRepositoryInterface */
     private $orderRepository;
 
     /** @var AvailabilityCheckerInterface  */
     private $availabilityChecker;
 
+    /** @var ProductVariantInterface|null */
+    private $productVariant;
+
+    /** @var string */
+    private $productVariantCode = '';
+
     public function __construct(
         ProductVariantRepositoryInterface $productVariantRepository,
+        RepositoryInterface $orderItemRepository,
         OrderRepositoryInterface $orderRepository,
         AvailabilityCheckerInterface $availabilityChecker
     ) {
         $this->productVariantRepository = $productVariantRepository;
+        $this->orderItemRepository = $orderItemRepository;
         $this->orderRepository = $orderRepository;
         $this->availabilityChecker = $availabilityChecker;
     }
 
     public function validate($value, Constraint $constraint)
     {
-        Assert::isInstanceOf($value, AddItemToCart::class);
+        if (!$value instanceof ChangeItemQuantityInCart && !$value instanceof AddItemToCart) {
+            throw new \InvalidArgumentException;
+        }
 
-        /** @var AddingEligibleProductVariantToCart $constraint */
-        Assert::isInstanceOf($constraint, AddingEligibleProductVariantToCart::class);
+        $this->assignProductVariantAndProductVariantCode($value);
 
-        /** @var ProductVariantInterface|null $productVariant */
-        $productVariant = $this->productVariantRepository->findOneBy(['code' =>$value->productVariantCode]);
+        /** @var EligibleProductVariantToCart $constraint */
+        Assert::isInstanceOf($constraint, EligibleProductVariantToCart::class);
 
-        if ($productVariant === null) {
+        if ($this->productVariant === null) {
             $this->context->addViolation(
                 $constraint->productVariantNotExistMessage,
```

Comment:
Also, is it really good validation message? We already had this variant in a cart and during the update it disappeared? No longer available seems like a better wording